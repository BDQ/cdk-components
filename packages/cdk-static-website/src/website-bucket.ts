import * as path from 'path';
import { CfnCloudFrontOriginAccessIdentity } from '@aws-cdk/aws-cloudfront';
import { Bucket } from '@aws-cdk/aws-s3';
import { BucketDeployment, Source } from '@aws-cdk/aws-s3-deployment';
import { Construct, RemovalPolicy } from '@aws-cdk/core';
import { CanonicalUserPrincipal  } from '@aws-cdk/aws-iam';

export interface WebsiteBucketProps {
  /** Name of the bucket */
  bucketName?: string;

  /** Policy to apply when the bucket is removed from this stack. */
  removalPolicy?: RemovalPolicy;

  source?: string;

  disableUpload?: boolean;

  /** The index page for the site like 'index.html' */
  websiteIndexDocument?: string;

  /** The error page for the site like 'error.html' */
  websiteErrorDocument?: string;
}

export class WebsiteBucket extends Construct {
  private readonly originId: CfnCloudFrontOriginAccessIdentity;
  private readonly bucket: Bucket;

  constructor(parent: Construct, name: string, props: WebsiteBucketProps) {
    super(parent, name);

    const {
      bucketName,
      removalPolicy,
      disableUpload = false,
      source,
      websiteIndexDocument,
      websiteErrorDocument
    } = props;

    this.bucket = new Bucket(this, 'WebsiteBucket', {
      bucketName,
      removalPolicy,
      websiteIndexDocument: websiteIndexDocument || 'index.html',
      websiteErrorDocument: websiteErrorDocument || 'error.html'
    });

    this.originId = new CfnCloudFrontOriginAccessIdentity(
      this,
      'OriginAccessIdentity',
      {
        cloudFrontOriginAccessIdentityConfig: {
          comment: `CloudFront OriginAccessIdentity for ${
            this.bucket.bucketName
          }`
        }
      }
    );

    this.bucket.grantRead(new CanonicalUserPrincipal(this.originId.attrS3CanonicalUserId))

    if (!disableUpload) {
      const placeHolderSource = path.join(__dirname, '..', 'website');

      new BucketDeployment(this, 'DeployWebsite', {
        source: Source.asset(source || placeHolderSource),
        destinationBucket: this.bucket
      });
    }
  }

  public get s3OriginSource() {
    return {
      originAccessIdentity: this.originId,
      s3BucketSource: this.bucket
    };
  }
}
