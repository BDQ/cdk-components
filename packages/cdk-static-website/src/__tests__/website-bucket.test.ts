import { Stack } from '@aws-cdk/core';
import { WebsiteBucket } from '../website-bucket';
import { toMatchCdkSnapshot } from '@cloudcomponents/jest-cdk-snapshot';

expect.extend({ toMatchCdkSnapshot });

describe('cdk-static-website: website-bucket', () => {
  it('snapshot', () => {
    const stack = new Stack();

    new WebsiteBucket(stack, 'WebsiteBucket', {
      bucketName: 'testbucket',
      disableUpload: true
    });

    expect(stack).toMatchCdkSnapshot();
  });
});
