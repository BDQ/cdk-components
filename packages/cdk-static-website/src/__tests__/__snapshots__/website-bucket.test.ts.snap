// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-static-website: website-bucket snapshot 1`] = `
Object {
  "Resources": Object {
    "WebsiteBucket3C288C20": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketName": "testbucket",
        "WebsiteConfiguration": Object {
          "ErrorDocument": "error.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "WebsiteBucketOriginAccessIdentity717BE380": Object {
      "Properties": Object {
        "CloudFrontOriginAccessIdentityConfig": Object {
          "Comment": Object {
            "Fn::Join": Array [
              "",
              Array [
                "CloudFront OriginAccessIdentity for ",
                Object {
                  "Ref": "WebsiteBucket3C288C20",
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "WebsiteBucketPolicy65FBF6CB": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "WebsiteBucket3C288C20",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "CanonicalUser": Object {
                  "Fn::GetAtt": Array [
                    "WebsiteBucketOriginAccessIdentity717BE380",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "WebsiteBucket3C288C20",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "WebsiteBucket3C288C20",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
